### ISSUE TITLE
Large schema cache reloads cause 100% CPU spike when schema cache metadata is accessed.

### ISSUE THREAD

**Issue by colophonemes** [PostgREST/postgrest#3046]
### Environment

* PostgreSQL version: 13.10 (also tested 13.11)
* PostgREST version: Known issue since v9.0, recently tested in 11.1.0 and 11.2.1
* Operating system: MacOS, Docker (ECS)

### Description of issue

Our **production** instance of PostgREST exhibits the following behaviour:
- Schema cache is reloaded (either via app restart or `NOTIFY pgrst reload`)
- Things work as normal for a few minutes
- Something actually needs to look at the updated schema cache (the most reliable way to reproduce is to load the OpenAPI spec from the `/` endpoint, but the behaviour seems to eventually happen regardless)
- PostgREST starts spinning CPU, pegging at 100% for about five mins
	- If you triggered the spike by attempting to load the OpenAPI schema endpoint, we at first get a partial response (always 135,063 bytes in my testing) that loads within a few seconds, with a truncated form of the JSON (it's not a valid JSON object that contains fewer entries, but a longer response that has been truncated). Eventually this response will time out
	- Subsequent reloads of the OpenAPI schema endpoint trigger a CPU spike of similar duration
	- There's usually two spikes (one per OpenAPI refresh), and then by the end of the second one, the OpenAPI schema request finishes buffering and the full JSON loads; occasionally there'll be a third spike that's a bit shorter than the previous two, and the OpenAPI schema resolves after that one.
	- Once the OpenAPI schema request has loaded fully, subsequent requests return ~immediately, until such time as the schema cache reloads again.

Here's me reproducing the issue in a local Docker container, connected to our production DB. The time it takes for the CPU spike to calm down is shorter, presumably because my laptop is beefier than the ECS instances we run our production PostgREST on.

![image](https://github.com/PostgREST/postgrest/assets/6322017/c82b7fbd-f0cb-45c1-aafe-fc8d9b481e22)

We've observed this issue since at least v9.0.0 (though possibly it was present earlier), through to the present day. I've just run through the testing steps above on both v11.1.0 and v11.2.1, and the behaviour is there in both.

#### Possibly an interaction with TimescaleDB?

My guess about why this happens in production but not dev is that we use [TimescaleDB](https://docs.timescale.com/) for a number of our tables that contain time-series data. This means that the number of tables in our schema changes dynamically as more data is added to parent tables, and Timescale chunks these into child tables. The child tables are stored on a `_timescaledb_internal` schema, e.g. `_timescaledb_internal._hyper_1_12345_chunk`. Some of our larger tables have upwards of 500 of these chunks attached.

Formatting and running the [tablesSqlQuery from PostgREST/SchemaCache.hs](https://github.com/PostgREST/postgrest/blob/f10d4139feb8ff2cd619126a14bc4df559e36e25/src/PostgREST/SchemaCache.hs#L586-L807) results in ~350 entries against a dev database, and **nearly 13,000** entries against prod. If I add `AND n.nspname !~ 'timescale'` to the end of that query then the prod query drops back to around 300 entries.

So, our guess is that our production PostgREST instances are choking on this massive amount of data when processing the schema cache (looking at the network graph in the image above, it looks like it's transferring about 10MB over the wire to reload the schema).

#### Possible mitigations

I guess that there are at least three possible mitigations that I can think of here:
- Exclude child tables from the schema cache (maybe gated under a config var; not sure if this has implications for resource embedding/FK detection)
- Special-case `_timescaledb_internal` and other Timescale schemas so they don't appear in the tables query (seems a bit gross, but maybe routes around needing to handle edge cases for the more general case of child tables)
- Add a config variable that allows us to exclude specified schemas from the schema cache (probably safer + more general)

Obviously all these mitigations rely on my being correct that the root cause of the issue is the large number of child tables present in the tables query.

**Comment by colophonemes** [comment]
> Is this timescale schema part of your search path, i.e. via db-extra-search-path?

Sorry, forgot to mention that in the original post – no, we only have `api` in `db-schemas`, and then the schemas that `api` is an overlay on in `db-extra-search-path` (`app_schema_1`, `app_schema_2` etc). 

The tables on e.g. `app_schema_1` etc do themselves have TimescaleDB child tables, but these aren't hosted in `app_schema_1`, but live on their own `_timescaledb_internal` schema.

```
\d+ app_schema_1.some_table

                                       Table "app_schema_1.some_table"
┌──────────┬──────────────────────────┬───────────┬──────────┬─────────┬─────────┬──────────────┬─────────────┐
│  Column  │           Type           │ Collation │ Nullable │ Default │ Storage │ Stats target │ Description │
├──────────┼──────────────────────────┼───────────┼──────────┼─────────┼─────────┼──────────────┼─────────────┤
│ ts       │ timestamp with time zone │           │ not null │         │ plain   │              │             │
│ thing_id │ bigint                   │           │ not null │         │ plain   │              │             │
│ value    │ numeric                  │           │ not null │         │ main    │              │             │
└──────────┴──────────────────────────┴───────────┴──────────┴─────────┴─────────┴──────────────┴─────────────┘
Indexes:
    "some_table_thing_id_ts_idx" btree (thing_id, ts DESC)
Check constraints:
    "positive_value" CHECK (value > 0::numeric)
Triggers:
    ts_insert_blocker BEFORE INSERT ON some_table FOR EACH ROW EXECUTE FUNCTION _timescaledb_functions.insert_blocker()
Child tables: _timescaledb_internal._hyper_1_1_chunk,
              _timescaledb_internal._hyper_1_2_chunk,
              _timescaledb_internal._hyper_1_3_chunk,
              _timescaledb_internal._hyper_1_4_chunk,
              _timescaledb_internal._hyper_1_5_chunk,
              _timescaledb_internal._hyper_1_6_chunk,
              _timescaledb_internal._hyper_1_7_chunk,
              ... etc etc
Access method: heap
```

**Comment by colophonemes** [comment]
@wolfgangwalther Hey - just wanted to check in to see if this is something that is likely to be worked on in an upcoming release of PostgREST? If not, how feasible do you think it would be for me as a relatively inexperienced Haskell developer to dive in an try to submit a PR fixing this (and if this is the route we decide to go, do you have a preference for how a fix is implemented)?

**Comment by steve-chavez** [comment]
Since https://github.com/PostgREST/postgrest/pull/3213, we now log some schema cache stats:

```
20/Feb/2024:18:27:24 -0500: Schema cache queried in 12.7 milliseconds
20/Feb/2024:18:27:24 -0500: Schema cache loaded 264 Relations, 221 Relationships, 141 Functions, 15 Domain Representations, 45 Media Type Handlers
```

It would be interesting to know the metrics for your case @colophonemes.

If there's a large number of relations, perhaps we should exclude `_timescaledb_internal` from the schema cache?

**Comment by colophonemes** [comment]
@steve-chavez Heres the output from when I load using that build:

```
21/Feb/2024:16:23:22 +0000: Starting PostgREST 12.1 (1a141c1)...
21/Feb/2024:16:23:22 +0000: Attempting to connect to the database...
21/Feb/2024:16:23:22 +0000: Successfully connected to PostgreSQL 13.13 (Ubuntu 13.13-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, 64-bit
21/Feb/2024:16:23:22 +0000: Listening on port 3000
21/Feb/2024:16:23:24 +0000: Listening for notifications on the pgrst channel
21/Feb/2024:16:23:24 +0000: Config reloaded
21/Feb/2024:16:24:04 +0000: Schema cache queried in 640.9 milliseconds
21/Feb/2024:16:24:04 +0000: Schema cache loaded 99 Relations, 230 Relationships, 16 Functions, 0 Domain Representations, 4 Media Type Handlers
```
And here's some output from `ps` at 5s intervals while it's starting:

```sh
ELAPSED  %CPU %MEM
  00:01   2.0  0.2
  00:06   0.0  0.2
  00:11   1.6  0.3
  00:17   1.5  0.3
  00:22   1.1  0.4
  00:27   0.0  0.4
  00:32   0.0  0.4
  00:37   7.6  0.5
  00:42   1.3  0.5
# 21/Feb/2024:16:24:04 +0000: Schema cache queried in 640.9 milliseconds
  00:47 100.1  0.6
  00:52 100.1  0.6
  00:57  98.0  0.6
  01:02  98.2  0.6
  01:07 100.1  0.6
  01:12  99.1  0.6
  01:17 100.1  0.6
  01:22 100.1  0.6
  01:27 100.1  0.6
  01:32  99.1  0.6
  01:37 100.1  0.6
  01:42 100.1  0.6
  01:47  98.7  0.6
  01:52 100.1  0.6
  01:57  99.1  0.6
  02:02 100.1  0.6
  02:07 100.1  0.6
  02:12 100.1  0.6
  02:17  99.7  0.6
  02:22  98.9  0.6
  02:27  98.1  0.6
  02:32 100.1  0.6
  02:37 100.1  0.6
  02:42 100.1  0.6
  02:48 100.1  0.6
# 21/Feb/2024:16:24:04 +0000: Schema cache loaded 99 Relations, 230 Relationships, 16 Functions, 0 Domain Representations, 4 Media Type Handlers
  02:53   3.2  0.6
  02:58   0.0  0.6

```

**Comment by steve-chavez** [comment]
@colophonemes Aha. So that means the schema cache queries are not the problem since they're fast. The problem is the schema cache parsing. This was previously discussed on https://github.com/PostgREST/postgrest/issues/2450#issuecomment-1931254688.

**Comment by colophonemes** [comment]
Yeah, I agree that it seems like it's the parsing rather than the DB query (indeed, I was able to extract and manually run the SQL queries myself and they were very fast), but I think I agree with @wolfgangwalther that if we could cut down the number of relations we're feeding into the parsing step, then parsing would go a lot faster.

> If there's a large number of relations, perhaps we should exclude _timescaledb_internal from the schema cache?

Yeah, this seems like a straightforward solution, it could be a flag that's sort of the opposite of `db-extra-search-path`, giving a list of schemas that you want to exclude. But perhaps it's just better to go with the suggestion of cutting out more cruft from the tables queries directly? 🤷‍♂️

**Comment by steve-chavez** [comment]
@colophonemes Could you share your schema (DDL only) privately? (email on profile) Then I can work on improving the query or maybe the Haskell code.

**Comment by steve-chavez** [comment]
> And here's some output from ps at 5s intervals while it's starting:

That time now should be clearer on main branch due to https://github.com/PostgREST/postgrest/pull/3253

**Comment by colophonemes** [comment]
> Could you share your schema (DDL only) privately? (email on profile) Then I can work on improving the query or maybe the Haskell code.

Unfortunately I don't think that I can that easily. It's not super-straightforward to dump out the Timescale stuff in a way that's restorable on your end. I'll try to prepare a toy schema that results in the same behaviour.

**Comment by steve-chavez** [comment]
Did some digging using the `big_schema`:

```
$ PGRST_DB_SCHEMAS="apflora" postgrest-with-postgresql-15 -f test/io/big_schema.sql postgrest-run

28/Feb/2024:13:33:24 -0500: Starting PostgREST 12.1 (160caaf)...
...
28/Feb/2024:13:33:24 -0500: Schema cache loaded 326 Relations, 305 Relationships, 6 Functions, 0 Domain Representations, 4 Media Type Handlers
28/Feb/2024:13:33:42 -0500: Schema cache loaded in 17141.6 milliseconds
```

> I'm not exactly sure how, but one way to do this could be to merge the allViewsKeysDependencies and tablesSqlQuery queries - so that we will never have to "query views first, then pass them to the tables query".

The problem is not related to the `tablesSqlQuery`, which is fast. It's only isolated to the relationships

https://github.com/PostgREST/postgrest/blob/97cd559265a22603b029893533524d8e44618f9a/src/PostgREST/SchemaCache.hs#L167

By not computing the relationships and just replacing it with a `mempty` the times gets reduced to:

```
28/Feb/2024:13:37:07 -0500: Starting PostgREST 12.1 (160caaf)...

28/Feb/2024:13:37:07 -0500: Schema cache loaded 326 Relations, 0 Relationships, 6 Functions, 0 Domain Representations, 4 Media Type Handlers
28/Feb/2024:13:37:07 -0500: Schema cache loaded in 1.4 milliseconds
```

Some profiling also indicates that it's not the query but the different tranformations we do on Haskell code that take a lot of time and CPU:

https://github.com/PostgREST/postgrest/blob/97cd559265a22603b029893533524d8e44618f9a/src/PostgREST/SchemaCache.hs#L163


It should be possible to do that logic on the db instead but it will take a fair amount of work.

**Comment by steve-chavez** [comment]
@colophonemes We've made significant improvements for the schema cache loading on https://github.com/PostgREST/postgrest/releases/tag/v14.0

> - Fix schema cache loading taking a long time for large schemas by @mkleczek in #4360, #3704
>    + On a complex database, [it goes down from 7 minutes to 2 seconds ](https://github.com/PostgREST/postgrest/pull/4396#issuecomment-3393243374).

Can you try and see if it solves your issue?

**Comment by colophonemes** [comment]
> Can you try and see if it solves your issue?

Thanks very much @steve-chavez, yep, this seems to be much better. My quick testing shows us loading the full schema cache in ~25 seconds, which is a huge improvement.

The times associated with schema cache querying/loading shown in the log don't seem to bear any resemblance to the actual loading time (note the ~23s gap between `Config reloaded` and `Schema cache queried...`), but I'm guessing this could just be how long it takes to pull schema information over the wire?

```
28/Oct/2025:16:01:29 +0000: Starting PostgREST 14.0...
28/Oct/2025:16:01:29 +0000: Admin server listening on 0.0.0.0:3011
28/Oct/2025:16:01:29 +0000: API server listening on 0.0.0.0:3002
28/Oct/2025:16:01:30 +0000: Successfully connected to PostgreSQL 13.13 (Ubuntu 13.13-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, 64-bit
28/Oct/2025:16:01:30 +0000: Connection Pool initialized with a maximum size of 10 connections
28/Oct/2025:16:01:30 +0000: Listening for database notifications on the "pgrst" channel
28/Oct/2025:16:01:31 +0000: Config reloaded
28/Oct/2025:16:01:54 +0000: Schema cache queried in 911.1 milliseconds
28/Oct/2025:16:01:54 +0000: Schema cache loaded 106 Relations, 267 Relationships, 19 Functions, 0 Domain Representations, 4 Media Type Handlers, 1224 Timezones
28/Oct/2025:16:01:54 +0000: Schema cache loaded in 412.1 milliseconds
```

Thanks for all your work here!

**Comment by mkleczek** [comment]
@colophonemes 
I'm somewhat suspicious 23s gap is caused by some external factor. CPU throttling? Network lags (it's always DNS, remember?).
The schema cache size is quite small (106 relations, 267 relationships), so I don't expect schema querying or loading (even before optimization) to be that slow.

Times logged seem fine to me as well. There must be something happening between config reloading and the start of schema querying.

**Comment by colophonemes** [comment]
Yeah, it's a strange one. Extracting the [`tablesSqlQuery`](https://github.com/PostgREST/postgrest/blob/v14.0/src/PostgREST/SchemaCache.hs#L575) query and running with `\timing` and `EXPLAIN ANALYZE` shows that it takes ~3s just to run the query and an additional 1s to read the data over the wire to my testing machine. So, that leaves about 19s to explain. The query does return 25k relations though, so I'm not sure how much of a bearing that has on things.

_ETA: I ran the query substituting in various combinations of my main API schema plus the additional schemas for the positional parameter, and the query always seemed to return ~25k rows_

_ETA2: I also don't think it's CPU issues – I'm running compiled from source with `stack build` natively on an M1 MacBook Pro, so no virtualization layer etc`_

**Comment by mkleczek** [comment]
> Yeah, it's a strange one. Extracting the [`tablesSqlQuery`](https://github.com/PostgREST/postgrest/blob/v14.0/src/PostgREST/SchemaCache.hs#L575) query and running with `\timing` and `EXPLAIN ANALYZE` shows that it takes ~3s just to run the query and an additional 1s to read the data over the wire to my testing machine. So, that leaves about 19s to explain. The query does return 25k relations though, so I'm not sure how much of a bearing that has on things.

Hmm... in the log dump you sent earlier we have:
```
28/Oct/2025:16:01:54 +0000: Schema cache loaded 106 Relations, 267 Relationships, 19 Functions, 0 Domain Representations, 4 Media Type Handlers, 1224 Timezones
28/Oct/2025:16:01:54 +0000: Schema cache loaded in 412.1 milliseconds
```

That's way less than 25k relations you get from running the query.

I'm wondering what your db connection settings are - do they point to many hosts?

**Comment by colophonemes** [comment]
> That's way less than 25k relations you get from running the query.

Yeah, per the original issue description, most of these are child tables created by TimescaleDB ('hypertables'). So, even though there are thousands of them in theory, there's only one table in the target schemas.

> I'm wondering what your db connection settings are - do they point to many hosts?

No, just a single host. I'm connecting via a Wireguard tunnel (Tailscale), but that's the same whether I'm running the query manually or passing connection options to PostgREST

**Comment by wolfgangwalther** [comment]
Seems like we need to filter out those child tables instead of returning them?

**Comment by colophonemes** [comment]
Yeah, ideally some way of excluding them would be good. My [previous suggestion](https://github.com/PostgREST/postgrest/issues/3046#issuecomment-1957517280) for this was basically to provide the ability to exclude schemas:

> Yeah, this seems like a straightforward solution, it could be a flag that's sort of the opposite of `db-extra-search-path`, giving a list of schemas that you want to exclude. But perhaps it's just better to go with the suggestion of cutting out more cruft from the tables queries directly? 🤷‍♂️

Not sure if that seems like a good idea, but could be an easy way out of this?

---

_Edit: deleted some debug information, I think it was spurious_

**Comment by steve-chavez** [comment]
>  I'm running compiled from source with stack build natively on an M1 MacBook Pro, so no virtualization layer etc`

@colophonemes Since you're building from source, could you try excluding the timescaledb schemas from the following query?

https://github.com/PostgREST/postgrest/blob/bccf7ac97a8a4d821fe199cab530ec3d37c6c653/src/PostgREST/SchemaCache.hs#L712

This is the list of timescale schemas: https://github.com/timescale/timescaledb/blob/main/sql/pre_install/schemas.sql

Then recheck the schema cache time.

If those are the culprit maybe we can just hardcode this case instead of having an additional config.

**Comment by steve-chavez** [comment]
> Yeah, it's a strange one. Extracting the [tablesSqlQuery](https://github.com/PostgREST/postgrest/blob/v14.0/src/PostgREST/SchemaCache.hs#L575) query and running with \timing and EXPLAIN ANALYZE shows that it takes ~3s just to run the query and an additional 1s to read the data over the wire to my testing machine. So, that leaves about 19s to explain. The query does return 25k relations though, so I'm not sure how much of a bearing that has on things.

@colophonemes We just added new logs that show exactly which schema cache query is slow, looks like this:

```
...
10/Apr/2026:21:48:45 -0500: Schema cache queried in 192.2 milliseconds
10/Apr/2026:21:48:45 -0500: tables: 72.027 ms, keydeps: 20.118 ms, rels: 6.189 ms, funcs: 35.010 ms, comprels: 4.319 ms, dreps: 1.614 ms, mhandlers: 7.419 ms, tzones: 43.025 ms
```

You  can try it using the latest devel release on https://github.com/PostgREST/postgrest/releases/tag/untagged-636540d9448009d20c7b with `log-level=debug`.
